Added compatibility with symfony 4

rvntone opened the pull request:
The Symfony\Component\Process\ProcessUtils::escapeArgument() method is deprecated since version 3.3 and will be removed in 4.0.

https://github.com/symfony/symfony/blob/master/UPGRADE-4.0.md#process

but the \Symfony\Component\Process\Process class has a private escapeArgument functions that you can used by calling the getCommandLine after a creation of the object.

ping @marcj

marcj commented on src/ProcessManager.php:
```diff
@@ -902,10 +902,16 @@ protected function newSlaveInstance($port)
+            $commandline = ['exec', $this->phpCgiExecutable, '-C', $file];
+            $process = new \Symfony\Component\Process\Process($commandline);
+            $commandline = $process->getCommandLine();
+        }
```
Well, we could simply go the way with an array all the time, can't we? Or does older Symfony not support arrays as first argument for `Process`?

```
$commandline = ['exec', $this->phpCgiExecutable, '-C', $file];
// use exec to omit wrapping shell	         // use exec to omit wrapping shell
$process = new Process('exec ' . $commandline);	+        $process = new Process($commandline);

`

rvntone commented:
I made that change, but I checked and in the vesion 2.8 of the Process
pacakge and it doesn't have the escaping argument thing, and it doesn't
support the array parameter.

The problem with the method_exists function is that there is a bug in the
phpStan that give you a false positive and the travis tests fail. (
https://github.com/phpstan/phpstan/issues/804)

The other thing that it can be done is remove the support of the 2.x
versions.



On Sat, Mar 3, 2018 at 2:14 PM, Marc J. Schmidt <notifications@github.com>
wrote:

> *@marcj* commented on this pull request.
> ------------------------------
>
> In src/ProcessManager.php
> <https://github.com/php-pm/php-pm/pull/360#discussion_r172025699>:
>
> > @@ -902,10 +902,16 @@ protected function newSlaveInstance($port)
>          // we can not use -q since this disables basically all header support
>          // but since this is necessary at least in Symfony we can not use it.
>          // e.g. headers_sent() returns always true, although wrong.
> -        $commandline = $this->phpCgiExecutable . ' -C ' . ProcessUtils::escapeArgument($file);
> +        if (method_exists('Symfony\Component\Process\ProcessUtils', 'escapeArgument')) {
> +            $commandline = 'exec ' . $this->phpCgiExecutable . ' -C ' . ProcessUtils::escapeArgument($file);
> +        } else {
> +            $commandline = ['exec', $this->phpCgiExecutable, '-C', $file];
> +            $process = new \Symfony\Component\Process\Process($commandline);
> +            $commandline = $process->getCommandLine();
> +        }
>
> Well, we could simply go the way with an array all the time, can't we? Or
> does older Symfony not support arrays as first argument for Process?
>
> $commandline = ['exec', $this->phpCgiExecutable, '-C', $file];
> // use exec to omit wrapping shell	         // use exec to omit wrapping shell
> $process = new Process('exec ' . $commandline);	+        $process = new Process($commandline);
>
> `
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/php-pm/php-pm/pull/360#pullrequestreview-100983894>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACnsH0z4IxCeN-h7q-DGtUcaj9B2FaRlks5tauuDgaJpZM4SbBL0>
> .
>



-- 
Yonel Meza Avila

rvntone commented:
ping @marcj
I added the if method_exists again and added a line to ignore the phpStan error in the ".phpstan.neon" file, I think that the method_exists is the right approach, but due the phpstan bug the line in the ".phpstan.neon" is needed.

andig commented:
Lgtm. The if is necessary since older symfony doesn‘t accept arrays as cmdline. 

Ping @marcj

andreybolonin commented:
any news when this or my pr would be merged?

acasademont commented:
Could this be merged soon please? Thanks <3

andig commented:
Given that @marcj appears to be busy I'm merging without further feedback.

acasademont commented:
Thanks a lot!!

andreybolonin commented:
@andig what about to tag new release?
https://github.com/php-pm/php-pm-httpkernel/issues/97

marcj commented:
Thanks!
